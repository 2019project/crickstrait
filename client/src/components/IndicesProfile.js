import React, { Component } from "react";
import { getIndicesById, getOhlcChartIndex } from "../actions/Indices";
import { connect } from "react-redux";
import "../styles/IndicesProfile.css";
import Loader from "react-loader-spinner";

export class IndicesProfile extends Component {
  componentDidMount() {
    this.props.getIndicesById(this.props.match.params.id);
    this.props.getOhlcChartIndex(this.props.match.params.id);
  }
  render() {
    console.log(this.props.sector);
    return (
      <div>
        <div id="indicesMainContainer">
          {this.props.singleIndex.map(index => (
            <>
              <div id="indicesLeftContainer">
                <h1>{index.ticker_name}</h1>
                <p>
                  Closed Price :{" "}
                  {Number(index.ticker_dates["2019-11-05"].closing).toFixed(2)}
                </p>
                {/* <hr /> */}
              </div>
              <div id="indicesGraphContainer">
                <h3>Stock Chart</h3>
                {/* {this.props.isLoading ? ( 
                  <div style={{ margin: "200px 500px" }}>
                    <Loader
                      type={Loader}
                      color="#2c3e50"
                      height="100"
                      width="400"
                    />
                    <img src={loader} alt="loading..." />
                  </div>
                ) : (
                  <div
                    style={{
                      border: "1px solid #cacaca",
                      width: "97%",
                      margin: "auto",
                      marginBottom: "20px"
                    }}
                  >
                    <iframe
                      src={this.props.ohlc_chart}
                      style={{
                        width: "100%",
                        height: "550px",
                        outline: "none",
                        border: "none"
                      }}
                    />
                  </div>
                )} */}
              </div>
              <div id="losersGainersContainer">
                <h3>Historic Prices</h3>
                <h4>Date : </h4>
                <table id="historicTable">
                  <tr>
                    <th>Date</th>
                    <th>Open</th>
                    <th>High</th>
                    <th>Low</th>
                    <th>Close</th>
                    <th>Adj Close</th>
                    <th>Volume</th>
                  </tr>
                </table>
              </div>
            </>
          ))}
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  singleIndex: state.indexReducer.singleIndex,
  ohlcChart: state.indexReducer.ohlcChart,
  isLoading: state.LoadingReducer.isLoading
});
export default connect(mapStateToProps, {
  getIndicesById,
  getOhlcChartIndex
})(IndicesProfile);
