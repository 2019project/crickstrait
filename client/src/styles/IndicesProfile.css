@import url(https://fonts.googleapis.com/css?family=Karla);
@import url("https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css");
* {
  font-family: Karla;
}

#indicesMainContainer {
  width: 100%;
  margin: 5px auto;
  display: grid;
  grid-template-rows: 130px auto auto auto 550px;
  grid-template-columns: 1fr;
  grid-gap: 10px;
}

#indicesLeftContainer h1 {
  font-weight: 500;
}

#indicesGraphContainer {
  padding: 20px;
  width: 97%;
  box-shadow: 0px 4px 6px 0px #cacaca;
}
#indicesTopLeft {
  display: grid;
  grid-template-columns: 1fr 1fr;
}

#downloadButton {
  margin-top: 20px;
  float: right;
  color: #39abf7;
  padding: 0px 20px;
  border: 2px solid #39abf7;
  cursor: pointer;
  border-radius: 4px;
  background-color: transparent;
}

#downloadButton:hover {
  color: #fff;
  background-color: #39abf7;
}

#losersGainersContainer table {
  box-shadow: 0px 4px 6px 0px #cacaca;
}

#historictTable {
  table-layout: fixed;
  width: 100%;
  border-collapse: collapse;
}

#historicTable tbody {
  display: block;
  overflow: auto;
  height: 200px;
}

#historicTable thead tr {
  display: block;
}

#historicTable thead {
  background: #2c3e50;
  color: #fff;
}

#historicTable th,
#historicTable td {
  padding: 5px;
  text-align: left;
  width: 200px;
}

#indicesNav {
  overflow: hidden;
}

#indicesNavName {
  color: #707070;
  overflow: hidden;
  font-size: 15px;
}

#indexClose {
  float: left;
  text-align: center;
  font-size: 17px;
  font-weight: bold;
}

#indexOpen {
  float: right;
  padding-left: 20px;
  padding-right: 40px;
  text-align: center;
  font-size: 17px;
  font-weight: bold;
}

#indexOpenName {
  float: right;
  padding-left: 20px;
  padding-right: 70px;
  text-align: center;
  font-size: 12px;
}
#indexCloseName {
  float: left;
  text-align: center;
  font-size: 12px;
}

#losersGainersContainer h4 {
  padding: 10px;
  font-weight: bold;
}

#indicesHeaderLabel {
  color: #2c3e50;
  padding-left: 20px;
  margin: 0;
  margin-bottom: 10px;
  font-family: Karla;
  font-weight: bold;
}

#weekButton {
  background-color: transparent;
  color: #39abf7;
  padding: 8px;
  outline: none;
  border: 2px solid #39abf7;
  margin-left: 20px;
  border-radius: 2px;
  cursor: pointer;
}

#monthButton {
  background-color: transparent;
  color: #39abf7;
  padding: 8px;
  outline: none;
  border: 2px solid #39abf7;
  margin-left: 20px;
  border-radius: 2px;
  cursor: pointer;
}

#indicesButtonContainer {
  margin-bottom: 20px;
}

#weekButtonactive {
  background-color: #39abf7;
  color: #fff;
  margin-left: 20px;
  border-radius: 2px;
  cursor: pointer;
  padding: 8px;
  outline: none;
  border: 2px solid #39abf7;
}

#monthButtonactive {
  background-color: #39abf7;
  color: #fff;
  margin-left: 20px;
  border-radius: 2px;
  cursor: pointer;
  padding: 8px;
  outline: none;
  border: 2px solid #39abf7;
}

@media screen and (min-width: 430px) {
  #indicesMainContainer {
    width: 100%;
    margin: 5px auto;
    display: grid;
    grid-template-rows: 130px auto 550px;
    grid-template-columns: 1fr;
    grid-gap: 10px;
  }

  #indicesLeftContainer {
    padding: 0px 25px;
    font-size: 20px;
    color: #2c3e50;
    box-shadow: 0px 4px 6px 0px #cacaca;
  }

  #indicesGraphContainer {
    width: 97%;
    margin: auto;
  }

  #losersGainersContainer {
    width: 97%;
    margin: auto;
  }
}
