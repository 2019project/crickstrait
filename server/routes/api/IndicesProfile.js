const express = require("express");
const router = express.Router();
const stocksData = require("../../model/stocksModel");
//used to provide path for downloadable file[piyush]
const path = require("path");

router.get("/singleindex/:id", async (req, res, next) => {
  try {
    let id = req.params.id;
    let result = await stocksData.find({ ticker_id: +id });
    let change = [];
    result.forEach(function(elem) {
      let name = {};
      let result = Object.values(elem._doc.ticker_dates);
      last_date_value = result.slice(-1)[0];
      let ticker_name = elem._doc.ticker_name;
      let ticker_id = elem._doc.ticker_id;
      let closing = last_date_value.closing.toFixed(2);
      let opening = last_date_value.opening.toFixed(2);
      let adjclose = last_date_value.adjclose.toFixed(2);
      let high = last_date_value.high.toFixed(2);
      let low = last_date_value.low.toFixed(2);
      let percentChange = ((closing - opening) * 100) / opening;
      name["change_percent"] = percentChange.toFixed(2).toString();
      if (name["change_percent"].charAt(0) == "-") {
        flag = true;
      } else {
        flag = false;
        name["change_percent"] = "+" + name["change_percent"];
      }
      change.push(name);
      name["date"] = last_date_value.date;
      name["ticker_name"] = ticker_name;
      name["ticker_id"] = ticker_id;
      name["close"] = closing;
      name["open"] = opening;
      name["adjclose"] = adjclose;
      name["high"] = high;
      name["low"] = low;
    });
    if (result.length == 0) {
      res.status(400).json({
        status: 400,
        data: null,
        message: "No Indices Found"
      });
    } else {
      res.status(200).json({
        status: 200,
        data: change,
        message: "Retrieved Indices successfully"
      });
    }
  } catch (err) {
    next(err);
  }
});

//provides downloadable format according to the required market index[piyush]
router.get("/download/:index", async (req, res, next) => {
  const index = req.params.index;
  res.download(
    path.join(
      __dirname,
      "../../db-init/stock-data/yahoo-data/index/^" + index + ".csv"
    ),
    function(err) {
      console.log(err);
    }
  );
});

router.post("/ohlcdata/:id", async (req, res, next) => {
  try {
    let id = req.params.id;
    console.log(id);
    console.log(req.body.time);
    var index = [];
    if (req.body.time == "1w") {
      var result = await stocksData.aggregate([
        { $unwind: "$ticker_dates" },
        {
          $match: {
            ticker_id: +id,
            "ticker_dates.date": {
              $lte: new Date("2019-11-05"),
              $gte: new Date("2019-10-24")
            }
          }
        }
      ]);
    }

    if (req.body.time == "1m") {
      var result = await stocksData.aggregate([
        { $unwind: "$ticker_dates" },
        {
          $match: {
            ticker_id: +id,
            "ticker_dates.date": {
              $lte: new Date("2019-11-05"),
              $gte: new Date("2019-09-12")
            }
          }
        }
      ]);
    }
    result.forEach(function(elem) {
      let name = {};
      var ticker_dates = elem.ticker_dates;
      name["open"] = ticker_dates.opening.toFixed(2);
      name["high"] = ticker_dates.high.toFixed(2);
      name["low"] = ticker_dates.low.toFixed(2);
      name["close"] = ticker_dates.closing.toFixed(2);
      name["adjclose"] = ticker_dates.adjclose.toFixed(2);
      name["volume"] = ticker_dates.volume.toFixed(2);
      name["date"] = ticker_dates.date;
      index.push(name);
    });
    var index1 = [];
    index1 = index.reverse();
    if (result.length == 0) {
      res.status(400).json({
        status: 400,
        data: result,
        message: "No Data Found"
      });
    } else {
      res.status(200).json({
        status: 200,
        data: [index1],
        message: "Retrieved Data successfully"
      });
    }
  } catch (err) {
    next(err);
  }
});

module.exports = router;
