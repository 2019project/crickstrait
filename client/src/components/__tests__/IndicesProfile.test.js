import React from "react";
import { shallow, mount } from "enzyme";
import { IndicesProfile } from "../IndicesProfile";
const indicesProfile = jest.fn();
const singleIndex = [];
let match = { params: { id: 1 }, isExact: true, path: "", url: "" };
const getIndicesById = jest.fn();
const getOhlcChartIndex = jest.fn();
const getOhlcIndicesById = jest.fn();
const data = [
  [
    {
      date: "29-Nov-2019",
      open: 2546.0,
      high: 2546.0,
      low: 2546.0,
      close: 2546.0,
      adjclose: 2546.0,
      volume: 2546.0
    },
    {
      date: "29-Nov-2019",
      open: 2546.0,
      high: 2546.0,
      low: 2546.0,
      close: 2546.0,
      adjclose: 2546.0,
      volume: 2546.0
    }
  ]
];
console.log(data[0][0].date);
const wrapper = mount(
  <IndicesProfile
    indicesProfile={indicesProfile}
    singleIndex={singleIndex}
    match={match}
    getIndicesById={getIndicesById}
    getOhlcChartIndex={getOhlcChartIndex}
    getOhlcIndicesById={getOhlcIndicesById}
    data={data}
  />
);

describe("Testing IndicesProfile Component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("should mount the component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should check for component did mount as a function", () => {
    const componentDidMount = jest.spyOn(
      IndicesProfile.prototype,
      "componentDidMount"
    );
    wrapper.instance().componentDidMount();
    expect(componentDidMount).toHaveBeenCalled();
  });

  it("should check for presense of weekClick function", () => {
    let weekClick = jest.spyOn(wrapper.instance(), "weekClick");
    expect(weekClick).toBeTruthy();
  });

  it("should check for presense of monthClick function", () => {
    let monthClick = jest.spyOn(wrapper.instance(), "monthClick");
    expect(monthClick).toBeTruthy();
  });

  it("should have a container with name historicTable", () => {
    expect(wrapper.find("#historicTable")).toBeTruthy();
  });

  it("should have a container with name losersGainersContainer", () => {
    expect(wrapper.find("#losersGainersContainer")).toBeTruthy();
  });

  it("should have a date in each row ", () => {
    expect(wrapper.find("#ohlcdate0").props().children).toBe(data[0][0].date);
  });

  it("should have a container with name downloadButton", () => {
    wrapper.find("#downloadButton").simulate("click");
  });
});
