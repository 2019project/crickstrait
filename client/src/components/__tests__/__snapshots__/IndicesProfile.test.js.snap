// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing IndicesProfile Component should mount the component 1`] = `
<IndicesProfile
  data={
    Array [
      Array [
        Object {
          "adjclose": 2546,
          "close": 2546,
          "date": "29-Nov-2019",
          "high": 2546,
          "low": 2546,
          "open": 2546,
          "volume": 2546,
        },
        Object {
          "adjclose": 2546,
          "close": 2546,
          "date": "29-Nov-2019",
          "high": 2546,
          "low": 2546,
          "open": 2546,
          "volume": 2546,
        },
      ],
    ]
  }
  getIndicesById={
    [MockFunction] {
      "calls": Array [
        Array [
          1,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  getOhlcChartIndex={
    [MockFunction] {
      "calls": Array [
        Array [
          1,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  getOhlcIndicesById={
    [MockFunction] {
      "calls": Array [
        Array [
          1,
          Object {
            "time": "1w",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  indicesProfile={[MockFunction]}
  match={
    Object {
      "isExact": true,
      "params": Object {
        "id": 1,
      },
      "path": "",
      "url": "",
    }
  }
  singleIndex={Array []}
>
  <div>
    <div
      id="indicesMainContainer"
    />
  </div>
</IndicesProfile>
`;
